File: plugins/tauri-plugin-sqlite/android/src/main/java/SqlitePlugin.kt
package lsy969999.plugin.todolumi.sqlite

import android.app.Activity
import android.webkit.WebView
import app.tauri.Logger
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin
import app.tauri.plugin.Invoke
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@TauriPlugin
class SqlitePlugin(private val activity: Activity): Plugin(activity) {
    private lateinit var webView: WebView
    private lateinit var db: AppDatabase
    override fun load(webView: WebView) {
        super.load(webView);
        this.webView = webView;
        this.db = AppDatabase.getInstance(activity);
    }

    @Command
    fun ping(invoke: Invoke) {
        val args = invoke.parseArgs(PingArgs::class.java)
        invoke.resolveObject(mapOf("value" to args.value))
    }

    @Command
    fun getDbUserVersion(invoke: Invoke) {
        CoroutineScope(Dispatchers.IO).launch {
            val userVersion = AppDatabase.getDbUserVersion(activity)
            invoke.resolveObject(mapOf("version" to userVersion))
        }
    }

    @Command
    fun getAllTodo(invoke: Invoke) {
        CoroutineScope(Dispatchers.IO).launch {
            val todos = db.todoDao().fetchAllTodo()
            invoke.resolveObject(GetAllTodoRes(todos = todos))
        }
    }

    @Command
    fun insertTodo(invoke: Invoke) {
        val args = invoke.parseArgs(InsertTodoReq::class.java)
        CoroutineScope(Dispatchers.IO).launch {
            val todo = TbTodo(todo = args.todo.orEmpty())
            val rowid = db.todoDao().insertTodo(todo);
            invoke.resolveObject(InsertTodoRes(rowid = rowid))
        }
    }
}

File: plugins/tauri-plugin-sqlite/ios/Sources/ios/SqlitePlugin.swift
// The Swift Programming Language
// https://docs.swift.org/swift-book

import UIKit
import WebKit
import Tauri
import os.log
import SQLite3
import SQLite


class SqlitePlugin: Plugin {
    var webView: WKWebView!
    var db: Connection!
    @objc public override func load(webview: WKWebView) {
        self.webView = webview
        let fileURL = try! FileManager
            .default
            .url(for: .documentDirectory, in: .userDomainMask, appropriateFor: nil, create: true)
            .appendingPathComponent(db_name)
        self.db = try! Connection(fileURL.path)
        
        Logger.debug("migration before userVersion: \(String(describing: self.db.userVersion))")
        todolumi_migration(db: self.db)
        Logger.debug("migration after userVersion: \(String(describing: self.db.userVersion))")
    }
    @objc public func ping(_ invoke: Invoke) throws {
        Logger.debug("userVersion: \(String(describing: self.db.userVersion))")
        let args = try invoke.parseArgs(PingReq.self)
        invoke.resolve(PingRes(value: args.value))
    }
    @objc public func getDbUserVersion(_ invoke: Invoke) throws {
        let version = self.db.userVersion;
        invoke.resolve(GetDbUserVersionRes(version: version));
    }
    @objc public func getAllTodo(_ invoke: Invoke) throws {
        let todos = try TodoDao.getAllTodo(db: self.db);
        invoke.resolve(GetAllTodoRes(todos: todos));
    }
    @objc public func insertTodo(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(InsertTodoReq.self);
        let rowid = try TodoDao.insertTodo(db: self.db, todo: args.todo);
        invoke.resolve(InsertTodoRes(rowid: rowid));
    }
}

@_cdecl("init_plugin_sqlite")
func initPlugin() -> Plugin {
    return SqlitePlugin()
}
